### PR TITLE
Compound bubble approach for text messages

### PR DESCRIPTION
**Pull request by magic146**

### CODE REVIEW

File: ChattoAdditions/Source/Chat Items/CompoundMessage/Content/MessageContentFactoryProtocol.swift [hunk 1]
```diff
@@ -41,20 +43,26 @@ public protocol MessageContentFactoryProtocol {
...
 
 public final class AnyMessageContentFactory<Model>: MessageContentFactoryProtocol {
 
     private let _canCreateMessageModule: (Model) -> Bool
     private let _createMessageModule: (Model) -> MessageContentModule
     private let _createLayoutProvider: (Model) -> MessageManualLayoutProviderProtocol
+    private let _createMenuPresenter: (Model) -> ChatItemMenuPresenterProtocol?
 
     public init<U: MessageContentFactoryProtocol>(_ base: U) where U.Model == Model {
+        self.identifier = "\(type(of: base as Any))"
```
Comment (wiruzx) [inline ChattoAdditions/Source/Chat Items/CompoundMessage/Content/MessageContentFactoryProtocol.swift]: Can we add identifier to the protocol and move this implementation to the default implementation of the protocol?

Comment (magic146) [reply ChattoAdditions/Source/Chat Items/CompoundMessage/Content/MessageContentFactoryProtocol.swift]: Done.

File: ChattoAdditions/Source/Chat Items/CompoundMessage/Content/MessageContentFactoryProtocol.swift [hunk 2]
```diff
@@ -21,49 +21,94 @@
...
+    func createNewMessageView() -> UIView
+    func createContentPresenter(forModel model: Model) -> MessageContentPresenterProtocol
+    func unbindContentPresenter(_ presenter: MessageContentPresenterProtocol)
+    func bindContentPresenter(_ presenter: MessageContentPresenterProtocol, withView view: UIView, forModel model: Model)
     func createLayoutProvider(forModel model: Model) -> MessageManualLayoutProviderProtocol
+    func createMenuPresenter(forModel model: Model) -> ChatItemMenuPresenterProtocol?
+}
+
+public extension MessageContentFactoryProtocol {
+    var identifier: String { return "\(type(of: self as Any))" }
```
Comment (wiruzx) [inline ChattoAdditions/Source/Chat Items/CompoundMessage/Content/MessageContentFactoryProtocol.swift]: Could you try `String(describing: self)` instead?

File: ChattoAdditions/Source/Chat Items/CompoundMessage/Views/CompoundBubbleView.swift [hunk 3]
```diff
@@ -124,10 +124,21 @@ public final class CompoundBubbleView: UIView, MaximumLayoutWidthSpecificable, B
 
     private func updateViews() {
         guard let viewModel = self.viewModel, let style = self.style else { return }
-        self.borderImageView.image = style.borderImage(forViewModel: viewModel)
-        self.borderImageView.layer.mask = self.borderMaskLayer
-        let maskImage = style.maskingImage(forViewModel: viewModel)
-        self.layer.mask = UIImageView(image: maskImage).layer
+
+        if let borderImage = style.borderImage(forViewModel: viewModel) {
```
Comment (wiruzx) [inline ChattoAdditions/Source/Chat Items/CompoundMessage/Views/CompoundBubbleView.swift]:
```suggestion
        self.borderImageView.image = style.borderImage(forViewModel: viewModel)
```

File: ChattoAdditions/Source/Chat Items/CompoundMessage/CompoundMessagePresenter.swift [hunk 4]
```diff
@@ -67,28 +69,28 @@ public final class CompoundMessagePresenter<ViewModelBuilderT, InteractionHandle
...
+    open override class func registerCells(_ collectionView: UICollectionView) {
+        // Cell registration is happening lazily, right before the moment when a cell is dequeued.
     }
 
-    public override func dequeueCell(collectionView: UICollectionView, indexPath: IndexPath) -> UICollectionViewCell {
-        return collectionView.dequeueReusableCell(withReuseIdentifier: .compoundCellReuseId,
-                                                  for: indexPath)
+    open override func dequeueCell(collectionView: UICollectionView, indexPath: IndexPath) -> UICollectionViewCell {
+        let cellReuseIdentifier = self.compoundCellReuseId
+        collectionView.register(CompoundMessageCollectionViewCell<ModelT>.self, forCellWithReuseIdentifier: cellReuseIdentifier)
```
[the remark this answers is not in the document]
Comment (magic146) [reply ChattoAdditions/Source/Chat Items/CompoundMessage/CompoundMessagePresenter.swift]: Done.

File: ChattoAdditions/Source/Chat Items/CompoundMessage/Content/MessageContentFactoryProtocol.swift [hunk 5]
```diff
@@ -21,49 +21,94 @@
...
-public final class MessageContentModule {
-    public typealias Presenter = Any
-    public let view: UIView
-    public let showBorder: Bool
-    public let presenter: Presenter
+import Chatto
+
+public protocol MessageContentPresenterProtocol {
+
+    var showBorder: Bool { get }
```
[the remark this answers is not in the document]
Comment (magic146) [reply ChattoAdditions/Source/Chat Items/CompoundMessage/Content/MessageContentFactoryProtocol.swift]: Discussed. We will fix it later.

File: ChattoAdditions/Source/Chat Items/CompoundMessage/Content/MessageContentFactoryProtocol.swift [hunk 6]
```diff
@@ -21,49 +21,94 @@
...
+    public func contentWasTapped() {}
 }
 
 public protocol MessageContentFactoryProtocol {
     associatedtype Model
-    func canCreateMessageModule(forModel model: Model) -> Bool
-    func createMessageModule(forModel model: Model) -> MessageContentModule
+    var identifier: String { get }
+    func canCreateMessageContent(forModel model: Model) -> Bool
+    func createNewMessageView() -> UIView
```
[the remark this answers is not in the document]
Comment (magic146) [reply ChattoAdditions/Source/Chat Items/CompoundMessage/Content/MessageContentFactoryProtocol.swift]: Fixed.

File: ChattoAdditions/Source/Chat Items/CompoundMessage/Content/MessageContentFactoryProtocol.swift (hunk 2 above)
Comment (magic146) [reply ChattoAdditions/Source/Chat Items/CompoundMessage/Content/MessageContentFactoryProtocol.swift]: Replaced with `String(describing: type(of: self))` but I will have to doublecheck it works correctly without casting to `Any`.

**Comment by wiruzx** [comment]
Looks good to me

File: ChattoAdditions/Source/Chat Items/CompoundMessage/CompoundMessagePresenter.swift [hunk 7]
```diff
@@ -67,60 +70,96 @@ public final class CompoundMessagePresenter<ViewModelBuilderT, InteractionHandle
...
-        return collectionView.dequeueReusableCell(withReuseIdentifier: .compoundCellReuseId,
-                                                  for: indexPath)
+    open override func dequeueCell(collectionView: UICollectionView, indexPath: IndexPath) -> UICollectionViewCell {
+        let cellReuseIdentifier = self.compoundCellReuseId
+        collectionView.register(CompoundMessageCollectionViewCell.self, forCellWithReuseIdentifier: cellReuseIdentifier)
+        return collectionView.dequeueReusableCell(withReuseIdentifier: cellReuseIdentifier, for: indexPath)
     }
 
-    public override func heightForCell(maximumWidth width: CGFloat,
+    open override func heightForCell(maximumWidth width: CGFloat,
```
Comment (AntonPalich) [inline ChattoAdditions/Source/Chat Items/CompoundMessage/CompoundMessagePresenter.swift]: Indentations are not correct for some functions now after changing access modifier from `public` to `open`

File: ChattoAdditions/Source/Chat Items/CompoundMessage/CompoundMessagePresenter.swift [hunk 8]
```diff
@@ -67,60 +70,96 @@ public final class CompoundMessagePresenter<ViewModelBuilderT, InteractionHandle
...
-            let modules = sSelf.contentFactories.map { $0.createMessageModule(forModel: sSelf.messageModel) }
+
             let bubbleView = compoundCell.bubbleView!
             bubbleView.viewModel = sSelf.messageViewModel
-            bubbleView.style = sSelf.compoundCellStyle
-            bubbleView.decoratedContentViews = modules.map { .init(module: $0) }
             bubbleView.layoutProvider = sSelf.layoutProvider
-            bubbleView.accessibilityIdentifier = sSelf.accessibilityIdentifier
+            bubbleView.style = sSelf.compoundCellStyle
+            bubbleView.setNeedsLayout()
```
Comment (AntonPalich) [inline ChattoAdditions/Source/Chat Items/CompoundMessage/CompoundMessagePresenter.swift]: Why this call is needed here? Shouldn't it be the responsibility of each concrete view to update its own layout based on state changes?

File: ChattoAdditions/Source/Chat Items/CompoundMessage/CompoundMessagePresenter.swift (hunk 7 above)
Comment (magic146) [reply ChattoAdditions/Source/Chat Items/CompoundMessage/CompoundMessagePresenter.swift]: Fixed.

File: ChattoAdditions/Source/Chat Items/CompoundMessage/CompoundMessagePresenter.swift (hunk 8 above)
Comment (magic146) [reply ChattoAdditions/Source/Chat Items/CompoundMessage/CompoundMessagePresenter.swift]: Fixed.

**Pull request merged**